src/Chainweb/Pact/Backend/Utils.hs

```diff
@@ -0,0 +1,204 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Module: Chainweb.Pact.ChainwebPactDb
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Emmanuel Denloye-Ito <emmanuel@kadena.io>
+-- Stability: experimental
+--
+
+module Chainweb.Pact.Backend.Utils where
+
+import Control.Concurrent.MVar
+import Control.Exception hiding (try)
+import Control.Exception.Safe hiding (bracket)
+import Control.Lens
+import Control.Monad.State.Strict
+
+import Control.Monad.Reader
+
+import Data.Bits
+import Data.String
+import Data.String.Conv
+import Data.ByteString hiding (pack)
+import Data.Text (pack, Text)
+import Database.SQLite3.Direct as SQ3
+
+import Prelude hiding (log)
+
+import System.Directory (removeFile)
+import System.IO.Extra
+
+-- pact
+
+import Pact.Types.Persistence
+import Pact.Types.SQLite
+import Pact.Types.Term(KeySetName(..), NamespaceName(..), ModuleName(..), PactId(..))
+import Pact.Types.Util (AsString(..))
+
+-- chainweb
+
+import Chainweb.Pact.Backend.Types
+import Chainweb.Pact.Service.Types
+import Chainweb.Pact.Backend.SQLite.DirectV2
+
+runBlockEnv :: MVar (BlockEnv SQLiteEnv) -> BlockHandler SQLiteEnv a -> IO a
+runBlockEnv e m = modifyMVar e $
+  \(BlockEnv  dbenv bs)  -> do
+    (a,s) <- runStateT (runReaderT (runBlockHandler m) dbenv) bs
+    return (BlockEnv dbenv s, a)
+
+callDb :: (MonadCatch m, MonadReader (BlockDbEnv SQLiteEnv) m, MonadIO m) => Text -> (Database -> IO b) -> m b
+callDb callerName action = do
+  c <- view (bdbenvDb . sConn)
+  res <- tryAny $ liftIO $ action c
+  case res of
+    Left err -> internalError $ "callDb (" <> callerName <> "): " <> (pack $ show err)
+    Right r -> return r
+
+withSavepoint :: SavepointName -> BlockHandler SQLiteEnv a -> BlockHandler SQLiteEnv a
+withSavepoint name action = do
+  beginSavepoint name
+  result <- try action
+  case result of
+    Right r -> do
```

Comment:
add bang to `r` or change `return r` to `liftIO $ evaluate r`

Reply:
done